### PR TITLE
feat: add send to server in admin gui

### PR DESCRIPTION
**Pull request by gabrielchl**
### Brief description of the changes:
Add item in admin GUI special games tab to send players to external server
so that we don't need to find the command everytime

image

### Tested Changes:
- [ ] I've tested the changes locally or with people.
- [x] I've not tested the changes. (can't connect to the same server with the same account name)

### CODE REVIEW

**Review by TheBigZet — CHANGES_REQUESTED** [review]
Add automatic popup for joining players.
Use `Event.add_removable()` with `Task.set_timeout()` and `Event.remove_removable()` so that it's active for, let's say, just an hour.

**Comment by gabrielchl** [comment]
> Add automatic popup for joining players. Use `Event.add_removable()` with `Task.set_timeout()` and `Event.remove_removable()` so that it's active for, let's say, just an hour.

Added the event handler. For unregistering the event handler, I opted for a command instead of a time-based approach as we rarely know how long a game lasts.
(untested btw, I'll need someone to test it / test it with)

File: comfy_panel/special_games.lua [hunk 1]
```diff
@@ -503,13 +503,23 @@ local function on_gui_click(event)
...
 		end
+
+		global.send_to_external_server_player_joined_game_handler = function(event)
+			game.get_player(event.player_index).connect_to_server{
+				address,
+				name,
+				description,
+			}
+		end
+		Event.add_removable_function(defines.events.on_player_joined_game, global.send_to_external_server_player_joined_game_handler)
```
Comment (TheBigZet) [inline comfy_panel/special_games.lua:522]:
```suggestion
		Event.add_removable_function(defines.events.on_player_joined_game, global.send_to_external_server_player_joined_game_handler, "server_switch_popup")
```

File: comfy_panel/special_games.lua [hunk 2]
```diff
@@ -540,6 +550,17 @@ local function on_player_died(event)
...
 
+commands.add_command("stop-send-to-external-server", nil, function(command)
+	local command_player = game.get_player(command.player_index)
+	if not command_player.admin then
+		command_player("Only admins can stop sending players to external server")
+		return
+	end
+
+	command_player.print("Stopped sending players to external server")
+	Event.remove_removable_function(defines.events.on_player_joined_game, global.send_to_external_server_player_joined_game_handler)
```
Comment (TheBigZet) [inline comfy_panel/special_games.lua:561]:
```suggestion
	Event.remove_removable_function(defines.events.on_player_joined_game, "server_switch_popup")
```

**Review by TheBigZet — CHANGES_REQUESTED** [review]
Seems like `add_removable_function` requires 3 parameters : event, function and name. Then it's referred to with `name` in `remove_removable_function`.

Also, the function probably doesn't need to be in `global` table

**Comment by gabrielchl** [comment]
> Seems like `add_removable_function` requires 3 parameters : event, function and name. Then it's referred to with `name` in `remove_removable_function`.
> 
> Also, the function probably doesn't need to be in `global` table. \edit Oh and I'd also prefer to have some safety mechanism, as admins like to forget about stuff, like disabling the popup :)

this is the docs in the events file
I'll add the time triggered safety mechanism

Screenshot 2023-05-21 130958

**Comment by TheBigZet** [comment]
Huh, then that example is wrong. Check the function definition and the comment right before it

**Comment by TheBigZet** [comment]
@gabrielchl tested on test server. Popup for connected players works. Popup for newly connecting players doesn't. Desync after opening `/debug`

**Comment by TheBigZet** [comment]
@XVhc6A @AwesomePatrol @gsquaredxc What do you think of that vs simple Event with one enabled/disabled condition? (In terms of performance and clarity)

**Comment by Ragnarok77-factorio** [comment]
> @XVhc6A @AwesomePatrol @gsquaredxc What do you think of that vs simple Event with one enabled/disabled condition? (In terms of performance and clarity)

Performance I don't think it matters much as the event "on player joined" is rarely called, and the code being done here is very fast to execute too I think..... Did a profiler : 
image



For clarity, don't really care which implementation is done. I've just a doubt about how robust this removable function is though, gonna test something.

**Pull request merged**